Server socket impovements

s0me0ne-unkn0wn opened the pull request:
References https://github.com/sergot/openssl/pull/52

* Server context creation code is moved outside of supply{} block. Until we can handle several servers on a single port (that is, until SNI callback is implemented) we don't need a new context for every incoming connection and we definitely don't want to parse certificate files on every connection. 
* Added support for PKCS12 files containing server certificate, intermediate certificate chain and private key. Passphrase-protected PKCS12 are NYI.
* Exceptions are thrown if certificate/key files are specified, but cannot be opened/parsed/whatever.

I will add some tests later. For now, tested against OpenSSL 0.1.17 (with today's changes) on Ubutnu 16.04.3 (openssl 1.0.2g) and Win 7 (openssl 1.1.0g).

s0me0ne-unkn0wn commented:
@jnthn can you please issue a certificate signed by t/ca-crt.pem with this CSR, I don't want to re-generate all the chain to test an intermediate certificate.
Or you can share ca-crt's private key in repo, it may be useful later on.
[internim.zip](https://github.com/jnthn/p6-io-socket-async-ssl/files/1503434/internim.zip)

jnthn commented:
Sorry for the delay on this, had quite a bit on my plate. Will look at it during the next week.

jnthn commented:
@s0me0ne-unkn0wn Sorry for the delay, I've time to look at this now. Unfortunately, I didn't manage to track down the private key I used to make the test certs/keys. :-( So I think we'll need to regenerate a new one (and commit the private key used this time). If that's something you can take care of, please feel free to include new ones in the PR; otherwise, I can do it.

jnthn commented on lib/IO/Socket/Async/SSL.pm6:
```diff
@@ -17,6 +18,12 @@ sub SSL_CTX_load_verify_locations(OpenSSL::Ctx::SSL_CTX, Str, Str) returns int32
+sub PKCS12_parse(Pointer, Str, CArray[Pointer], CArray[Pointer], CArray[Pointer])
+    returns int32 is native(&gen-lib) {*}
+sub OPENSSL_add_all_algorithms_noconf() is native(&crypto-lib) {*}
+sub OpenSSL_add_all_algorithms() is native(&crypto-lib) {*}
```
These two add_all_algorithms functions appear to not be used.

s0me0ne-unkn0wn replied:
Okay, I'll take care of certificates then.
*_add_all_algorithms_* have to be used before PKCS12_parse() in try {} block, I remember doing that but cannot find in diff, probably I forgot to push something :/
I'll check that, thank you!

jnthn commented:
@s0me0ne-unkn0wn Quick update: I added various other things recently. Along the way, I applied a change like the one you suggested for moving context setup outside of the `whenever`; this was done in cf8778ba83a616442d1a64df1798d2740c5ba92e. However, I moved it just out of the `whenever`, but not out of the `supply` block, since that allows for getting the lifetime management of it more correct while still giving the same win.

s0me0ne-unkn0wn commented:
Okay, I hope that's it.
Keys and certificates are regenerated, tests are fixed to use new ones, I also added a simple bash script if it is ever needed to regenerate all over again.
Also removed `*_add_all_algoritms_*` import. Older openssl versions require to call one of those before importing PKCS#12 bundle (and so did perl5's IO::Socket::SSL), but it seems 1.0.2g does all the work needed in `SSL_library_init()`.  Anyway, it works for me without explicit call but I haven't run any tests on Windows.

jnthn commented:
Looks good to me, and the tests seem happy. Many thanks!
